src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,390 @@
+        )
+        process.wait()
+
+        pip_install_result: int = process.returncode
+        if process.stdout:
+            process_output: str = "\n".join([line.strip() for line in process.stdout])
+            _logger.debug(process_output)
+
+        if process.stderr:  # pragma: no cover
+            error_output: str = "\n".join([line.strip() for line in process.stderr])
+            _logger.warning(error_output)
+    except FileNotFoundError:
+        raise ModuleNotFoundError(
+            f"Pip not found. Please install pip in your environment or specify the path to your pip executable as "
+            f"'{PIP_ENVIRONMENT_VARIABLE}' environment variable and try again."
+        )
+
+    if pip_install_result is not None and pip_install_result != 0:
+        raise RuntimeError(f"Pip failed with return code {pip_install_result}.")
+
+
+def detect_native_dependencies(
+    target: str, downloaded_packages_dict: Dict[Requirement, List[str]]
+) -> Set[str]:
+    """
+    Detects files with native extensions present at the `target` folder, and deduces which packages own these files.
+    Native dependencies use C/C++ code that won't work when uploaded via a zip file. We detect these so that we can
+    switch to Anaconda-supported versions of these packages, where possible (or warn the user if it is not possible).
+
+    We detect native dependency by looking for file extensions that correspond to native code usage (Note that this
+    method is best-effort and will result in both false positives and negatives).
+
+    Args:
+        target (str): Target directory which contains packages installed by pip.
+        downloaded_packages_dict (Dict[Requirement, List[str]]): Mapping between packages and a list of files or
+        folders corresponding to it.
+
+    Returns:
+        Set[str]: Set of native dependency names. Note that we only return a set of strings here rather than Requirement
+        objects because the specific version of a native package is irrelevant.
+    """
+
+    def invert_downloaded_package_to_entry_map(
+        packages_dict: Dict[Requirement, List[str]]
+    ) -> Dict[str, Set[str]]:
+        """
+        Invert dictionary mapping packages to files/folders. We need this dictionary to be inverted because we first
+        discover files with native dependency extensions and then need to deduce the packages corresponding to these
+        files.
+
+        Args:
+            packages_dict (Dict[Requirement, List[str]]): Mapping between packages and a list of files or folders
+            corresponding to it.
+
+        Returns:
+            Dict[str, Set[str]]: The inverse mapping from a file or folder to the packages it corresponds to. Note that
+            it is unlikely a file corresponds to multiple packages (but we allow for the possibility). We only need
+            to return a set of strings here rather than Requirement objects because the specific version of a native
+            package is irrelevant.
```

Comment:
instead of using _corresponds to_, using _depend on it_ is better in this case.

Reply:
The package "is" those files and folders; what would be the right description